clang/lib/Sema/Sema.cpp

```diff
@@ -1386,6 +1386,13 @@ Sema::Diag(SourceLocation Loc, const PartialDiagnostic& PD) {
 static void emitCallStackNotes(Sema &S, FunctionDecl *FD) {
   auto FnIt = S.DeviceKnownEmittedFns.find(FD);
   while (FnIt != S.DeviceKnownEmittedFns.end()) {
+    if (S.getLangOpts().SYCLIsDevice &&
+        FnIt->second.FD->hasAttr<SYCLKernelAttr>()) {
+      // Skip over the routines with sycl_kernel attributes
+      // in the traceback as they are likely from SYCL headers.
+      FnIt = S.DeviceKnownEmittedFns.find(FnIt->second.FD);
```

Comment:
Isn't this setting the next loop to the thing you just made sure has the kernel attribute?  Shouldn't you be setting it to the 'next' one, or at least checking something different?  What is the contents of DeviceKnownEmittedFns?  Using a std::pair for that seems like a mistake here.

Reply:
No, what it does is find that this routine has the kernel attribute, and then tries to find its caller.
DeviceKnownEmittedFns is a map (llvm::DenseMap), with callee paired with its known caller and its location.

I am not very comfortable changing that data structure (used elsewhere too) and definitely don't want to change it as part of this change.